File: Assets/SideScroller/Scripts/SmoothFollow.cs
using UnityEngine;
using System.Collections;

public class SmoothFollow : MonoBehaviour {
	public Transform target;
	public float distance = 0.0f;
	public float height = 0.0f;
	public float damping;
	public bool smoothRotation = true;
	public float rotationDamping = 10.0f;
	
	void Update () {


		Vector3 wantedPosition = target.TransformPoint(0, height, -distance);
		//Vector3 wantedPosition = target.TransformPoint(-distance, height, 0);
		//transform.position = Vector3.Lerp (transform.position, wantedPosition, Time.deltaTime * damping);
		transform.position = Vector3.Lerp (transform.position, wantedPosition,  damping);

		transform.LookAt(target);
		/*
		if (smoothRotation) {
			Quaternion wantedRotation = Quaternion.LookRotation(target.position - transform.position, target.up);

			transform.rotation = Quaternion.Slerp (transform.rotation, wantedRotation, Time.deltaTime * rotationDamping);
		}
		
		else transform.LookAt (target, target.up);
		*/
	}
}

File: Assets/Drone Folder/Flight Scripts/SmoothFollow.js
/*
This camera smoothes out rotation around the y-axis and height.
Horizontal Distance to the target is always fixed.

There are many different ways to smooth the rotation but doing it this way gives you a lot of control over how the camera behaves.

For every of those smoothed values we calculate the wanted value and the current value.
Then we smooth it using the Lerp function.
Then we apply the smoothed values to the transform's position.
*/

// The target we are following
var target : Transform;
var target2 : Transform;
// The distance in the x-z plane to the target
var distance = 10.0;
// the height we want the camera to be above the target
var height = 5.0;
// How much we 
var heightDamping = 2.0;
var rotationDamping = 3.0;
var distanceOffset : float = 0.0;
// Place the script in the Camera-Control group in the component menu
@script AddComponentMenu("Camera-Control/Smooth Follow")

function Update()
{
    var relativePos = transform.position - (target2.position);
    var hit : RaycastHit;
    if (Physics.Raycast (target2.position, relativePos, hit, distance+0.5)) {
        Debug.DrawLine(target2.position,hit.point);
        distanceOffset = distance - hit.distance + 0.8; 
        distanceOffset = Mathf.Clamp(distanceOffset,0,distance);
    }
    //else
    //{
    //    distanceOffset = 0;
    //}
    
    if(distanceOffset >= 3)
    {
    	height = 0.1;
    }
    else if(distanceOffset <= 1.0)
    {
    	height = 1.0;
    }
}


function LateUpdate () 
{
	// Early out if we don't have a target
	if (!target)
		return;
	
	// Calculate the current rotation angles
	var wantedRotationAngle = target.eulerAngles.y;
	var wantedHeight = target.position.y + height;
		
	var currentRotationAngle = transform.eulerAngles.y;
	var currentHeight = transform.position.y;
	
	// Damp the rotation around the y-axis
	currentRotationAngle = Mathf.LerpAngle (currentRotationAngle, wantedRotationAngle, rotationDamping * Time.deltaTime);

	// Damp the height
	currentHeight = Mathf.Lerp (currentHeight, wantedHeight, heightDamping * Time.deltaTime);

	// Convert the angle into a rotation
	var currentRotation = Quaternion.Euler (0, currentRotationAngle, 0);
	
	// Set the position of the camera on the x-z plane to:
	// distance meters behind the target
	transform.position = target.position;
	transform.position -= currentRotation * Vector3.forward * distance;

	// Set the height of the camera
	transform.position.y = currentHeight;
	
	// Always look at the target
	transform.LookAt (target);
}